docs/designs/odata-writer-rewrite-proposal.md

```diff
@@ -0,0 +1,315 @@
+- New writer should expose similar levels of abstractions and extensibility to existing writer
+- New writer should provide support for all the scenarios that the existing writer supports
+  - Different input types (CLR objects, EdmObject, Dictionaries, JObject)
+  - Different types of payloads: resource set, values, aggregation results, service document, raw values, metadata document, etc.
+  - Dynamic properties, delta payloads, stream properties, polymorphic properties, $select/$expand
+  - Standard OData annotations, custom annotations, etc.
+  - Streaming support (start serialization before all data is available)
+
+The sample code in this doc or for demonstration purposes and are subject to change as the design is fleshed out more.
+
+## Top-level writer
+
+This is the equivalent to the existing `ODataMessageWriter` layer, which also exposes the abstract `ODataWriter` and `ODataOutputContext` classes.
+The `ODataMessageWriter` exposes methods for writing top-level payloads like metadata, service document, property values, etc. or methods
+for creating writers for writing more complex, nested payloads like resources and resouce sets.
+
+However, `ODataMessageWriter` doesn't implement any logic, all the logic is delegated to the abstract `ODataOutputContext` and `ODataWriter` classes.
+The `ODataOutputContext` does most of the heavy lifting and is the one that determines the output format.
+
+This means that if you wanted to create your own customer `ODataMessageWriter` from scratch (e.g. to output a different format other than JSON), you would
+need to provide custom `ODataOutputContext` and `ODataWriter` implementations. This is a non-trivial undertaking and I doubt there are any implementations
+in production other than those provided by the library.
+
+If you do create custom implementations of `ODataWriter` and `ODataOutputContext`, it is up to you to ensure that you adhere to OData's protocol, for example
+it's up to you to ensure the format of the response JSON matches the protocol, it is up to you to ensure you place
+
+The library's default implementations of `ODataWriter` and `ODataOutputContext` (e.g. `ODataJsonOutputContext`) are internal and not exposed to the caller,
+therefore they cannot be directly customized or extended. Instead OData provides extensions and customizability through the following mechanisms:
+
+- You can configure various settings and options in `ODataMessageWriterSettings`
+- You can replace the low-level `IJsonWriter` by implementing the interface
+- You can control what gets written through the properties and annotatons that you add to `ODataResource` and similar types
+
+Here's a rough proposal for the new writer:
+
+```csharp
+interface IODataMessageWriter<TContext, TState>
+{
+    IODataWriter<TContext, TState, TValue> GetWriter<TValue>(ODataPayloadKind payloadKind);
+    // Alternatively we could have different `GetWriter` methods for different

```

Comment:
please do this; i think avoiding enums is really helpful for us to have visibility into breaking changes